crates/egui/src/viewport.rs

```diff
@@ -0,0 +1,680 @@
+    #[inline]
+    fn from(id: ViewportId) -> Self {
+        id.0
+    }
+}
+
+impl nohash_hasher::IsEnabled for ViewportId {}
+
+/// A fast hash set of [`ViewportId`].
+pub type ViewportIdSet = nohash_hasher::IntSet<ViewportId>;
+
+/// A fast hash map from [`ViewportId`] to `T`.
+pub type ViewportIdMap<T> = nohash_hasher::IntMap<ViewportId, T>;
+
+// ----------------------------------------------------------------------------
+
+/// A pair of [`ViewportId`], used to identify a viewport and its parent.
+#[derive(Debug, Hash, Clone, Copy, PartialEq, Eq)]
+#[cfg_attr(feature = "serde", derive(serde::Deserialize, serde::Serialize))]
+pub struct ViewportIdPair {
+    pub this: ViewportId,
+    pub parent: ViewportId,
+}
+
+impl Default for ViewportIdPair {
+    #[inline]
+    fn default() -> Self {
+        Self::ROOT
+    }
+}
+
+impl ViewportIdPair {
+    /// The `ViewportIdPair` of the root viewport, which is its own parent.
+    pub const ROOT: Self = Self {
+        this: ViewportId::ROOT,
+        parent: ViewportId::ROOT,
+    };
+}
+
+/// The user-code that shows the ui in the viewport, used for deferred viewports.
+pub type ViewportUiCallback = dyn Fn(&Context) + Sync + Send;
+
+/// Render the given viewport, calling the given ui callback.
+pub type ImmediateViewportRendererCallback =
+    dyn for<'a> Fn(&Context, ViewportBuilder, ViewportIdPair, Box<dyn FnOnce(&Context) + 'a>);
+
+/// Control the building of a new egui viewport (i.e. native window).
+///
+/// The fields are public, but you should use the builder pattern to set them,
+/// and that's where you'll find the documentation too.
+///
+/// Since egui is immediate mode, `ViewportBuilder` is accumulative in nature.
+/// Setting any option to `None` means "keep the current value",
+/// or "Use the default" if it is the first call.
+#[derive(PartialEq, Eq, Clone)]
+#[allow(clippy::option_option)]
+pub struct ViewportBuilder {
+    pub id: ViewportId,
+
+    /// The title of the vieweport.
+    /// `eframe` will use this as the title of the native window.
+    pub title: Option<String>,
+
+    /// This is wayland only. See [`Self::with_name`].
+    pub name: Option<(String, String)>,
+
+    pub position: Option<Option<Pos2>>,
+    pub inner_size: Option<Option<Vec2>>,
+    pub fullscreen: Option<bool>,
+    pub maximized: Option<bool>,
+    pub resizable: Option<bool>,
+    pub transparent: Option<bool>,
+    pub decorations: Option<bool>,
+    pub icon: Option<Option<Arc<ColorImage>>>,
+    pub active: Option<bool>,
+    pub visible: Option<bool>,
+    pub title_hidden: Option<bool>,
+    pub titlebar_transparent: Option<bool>,
+    pub fullsize_content_view: Option<bool>,
+    pub min_inner_size: Option<Option<Vec2>>,
+    pub max_inner_size: Option<Option<Vec2>>,
+    pub drag_and_drop: Option<bool>,
+
+    pub close_button: Option<bool>,
+    pub minimize_button: Option<bool>,
+    pub maximize_button: Option<bool>,
+
+    pub hittest: Option<bool>,
+}
```

Comment:
There seem to be some options missing in `ViewportBuilder` that are present in `eframe::NativeOptions`, or them having different names, e.g.:

| NativeOptions | ViewportBuilder |
| - | - |
| always_on_top | - |
| centered | - |
| decorat**ed** | decorat**ions** |
| drag_and_drop<strong>_support</strong> | drag_and_drop |
| fullsize_content | fullsize_content<strong>_view</strong> |
| ... | ... |

Are the missing options left out on purpose?

If not, would it be a good idea to share them via a struct `NativeWindowOptions` (this could be a bit challenging since for `NativeOptions` all of them are `T` and for `ViewportBuilder` they are `Option<T>`) or a macro? That way it would also be easier to synchronize and maintain the options in the future.

Reply:
I'd like to remove all the window/viewport-related options from `NativeOptions` and use a `ViewportBuilder` there, but this could perhaps wait for a later PR